### ISSUE TITLE
How to load compressed model with vllm?

### ISSUE THREAD

**Issue by IEI-mjx** [vllm-project/llm-compressor#870]
I utilized LLMCompressor to quantize our model using the FP8-dynamic recipe. The quantized model was successfully tested using the SparseAutoModelForCausalLM method.

![image](https://github.com/user-attachments/assets/6a228800-164e-4708-a154-29d0f62c100f)

However, when attempting to load this quantized model using vllm, I encountered a KeyError. Upon inspecting the params_dict within the weight_load method, I noticed that there was no entry for "weight_scale" in the weight names.

Could you please advise on how I should pass parameters such as quant_config to ensure that these weight names are correctly recognized?

**Comment by dsikka** [comment]
Hi @IEI-mjx do you mind sharing the config of the model that you're loading into vllm and a code-snippet of how you're loading it?

Thanks!

**Comment by IEI-mjx** [comment]
Sure,
the config,
```
{
  "_from_model_config": true,
  "_name_or_path": "/mnt/beegfs2/maojunxiong/Yuan_Models/Yuan2-M32-MOD",
  "architectures": [
    "YuanForCausalLM"
  ],
  "attention_projection_size": 4096,
  "attn_dropout": 0.0,
  "auto_map": {
    "AutoConfig": "configuration_yuan.YuanConfig",
    "AutoModelForCausalLM": "yuan_moe_hf_model.YuanForCausalLM"
  },
  "bos_token_id": 77185,
  "causal_mask": true,
  "compression_config": {
    "config_groups": {
      "group_0": {
        "input_activations": {
          "actorder": null,
          "block_structure": null,
          "dynamic": true,
          "group_size": null,
          "num_bits": 8,
          "observer": "memoryless",
          "observer_kwargs": {},
          "strategy": "token",
          "symmetric": true,
          "type": "float"
        },
        "output_activations": null,
        "targets": [
          "Linear"
        ],
        "weights": {
          "actorder": null,
          "block_structure": null,
          "dynamic": false,
          "group_size": null,
          "num_bits": 8,
          "observer": "minmax",
          "observer_kwargs": {},
          "strategy": "channel",
          "symmetric": true,
          "type": "float"
        }
      }
    },
    "format": "float-quantized",
    "global_compression_ratio": 1.3110720166684395,
    "ignore": [
      "model.layers.0.mlp.gate.query_key_value",
      "model.layers.1.mlp.gate.query_key_value",
      "model.layers.2.mlp.gate.query_key_value",
      "model.layers.3.mlp.gate.query_key_value",
      "model.layers.4.mlp.gate.query_key_value",
      "model.layers.5.mlp.gate.query_key_value",
      "model.layers.6.mlp.gate.query_key_value",
      "model.layers.7.mlp.gate.query_key_value",
      "model.layers.8.mlp.gate.query_key_value",
      "model.layers.9.mlp.gate.query_key_value",
      "model.layers.10.mlp.gate.query_key_value",
      "model.layers.11.mlp.gate.query_key_value",
      "model.layers.12.mlp.gate.query_key_value",
      "model.layers.13.mlp.gate.query_key_value",
      "model.layers.14.mlp.gate.query_key_value",
      "model.layers.15.mlp.gate.query_key_value",
      "model.layers.16.mlp.gate.query_key_value",
      "model.layers.17.mlp.gate.query_key_value",
      "model.layers.18.mlp.gate.query_key_value",
      "model.layers.19.mlp.gate.query_key_value",
      "model.layers.20.mlp.gate.query_key_value",
      "model.layers.21.mlp.gate.query_key_value",
      "model.layers.22.mlp.gate.query_key_value",
      "model.layers.23.mlp.gate.query_key_value",
      "lm_head"
    ],
    "kv_cache_scheme": null,
    "quant_method": "compressed-tensors",
    "quantization_status": "compressed"
  },
  "dropout": 0,
  "eod_token": 77185,
  "eod_token_id": 77185,
  "eos_token_id": 77185,
  "hidden_act": "silu",
  "hidden_size": 2048,
  "initializer_range": 0.02,
  "intermediate_size": 8192,
  "lf_conv2d_group": 1,
  "lf_conv2d_num_pad": 0,
  "mask_token_id": 77185,
  "max_position_embeddings": 16384,
  "model_max_length": 8192,
  "model_type": "yuan",
  "moe_config": {
    "ffn_hidden_size": 8192,
    "gated_linear_unit": true,
    "moe_num_experts": 32,
    "moe_top_k": 2,
    "norm_topk_prob": true,
    "num_attention_router_heads": 8
  },
  "num_attention_heads": 16,
  "num_hidden_layers": 24,
  "output_router_logits": true,
  "pad_token_id": 77185,
  "reset_attention_mask": false,
  "reset_position_ids": true,
  "rms_norm_eps": 1e-06,
  "rotary_base": 40890,
  "rotary_percent": 1.0,
  "sep_token": 77187,
  "sep_token_id": 77185,
  "tie_word_embeddings": false,
  "tokenizer_class": "YuanTokenizer",
  "torch_dtype": "bfloat16",
  "transformers_version": "4.44.0",
  "use_cache": true,
  "use_flash_attention": true,
  "use_loss_mask": false,
  "use_moe": true,
  "vocab_size": 135040
}
```
And the code,
```
from vllm import LLM, SamplingParams
import time
from transformers import AutoTokenizer, LlamaTokenizer

def run_Yuan():
    tokenizer = LlamaTokenizer.from_pretrained("/mnt/beegfs2/maojunxiong/Yuan_Models/Yuan2-M32-MOD-FP8-Dynamic-LLMCompressor/", add_eos_token=False, add_bos_token=False, eos_token='<eod>', trust_remote_code=True)
    tokenizer.add_tokens(['<sep>', '<pad>', '<mask>', '<predict>', '<FIM_SUFFIX>', '<FIM_PREFIX>', '<FIM_MIDDLE>','<commit_before>','<commit_msg>','<commit_after>','<jupyter_start>','<jupyter_text>','<jupyter_code>','<jupyter_output>','<empty_output>'], special_tokens=True)
    sampling_params = SamplingParams(max_tokens=100, temperature=0.8, top_p=0.95, top_k=1)
    llm = LLM(model="/mnt/beegfs2/maojunxiong/Yuan_Models/Yuan2-M32-MOD-FP8-Dynamic-LLMCompressor/", tensor_parallel_size=1, disable_custom_all_reduce=True, max_num_seqs=1, enforce_eager=True, trust_remote_code=True, quantization='compressed-tensors')

    prompts = ['写一篇春游作文']
    outputs = []
    total_tokens = 0
    start_time = time.time()
    outputs = llm.generate(prompts, sampling_params)
    end_time = time.time()

    for output in outputs:
        prompt = output.prompt
        generated_text = output.outputs[0].text
        token = len(tokenizer.encode(generated_text))
        total_tokens += token
        print(f"Generated text: {generated_text!r}")

    print("inference_time:", (end_time - start_time))
    print("total_tokens:", total_tokens)

if __name__ == "__main__":
    run_Yuan()
```
vllm_v0.5.4 was used.

**Comment by dsikka** [comment]
@IEI-mjx  The parameter names are registered to each layer. For FP8, the following parameters are registered: 
https://github.com/vllm-project/vllm/blob/76ed5340f0ec0e481593ea1a94459b4b55136a4f/vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py#L101
Specifically, "weight" and "weight_scale". Therefore, the weight loading will expect to load these parameters from disk.

**Comment by IEI-mjx** [comment]
I have noticed this code as well. I am confused as to why the weight names printed from the line  `params_dict = dict(self.named_parameters())` do not include any `weight_scales` like this:
![image](https://github.com/user-attachments/assets/b1f80307-b38b-47af-b4ec-d28385dd807c)
Even though the quant_config is correctly generated from this line:
![image](https://github.com/user-attachments/assets/bb6050dd-166c-4071-958d-170371a15978)

**Comment by dsikka** [comment]
@IEI-mjx Can you share your model definition? The model should be defined using the LinearBase layers: https://github.com/vllm-project/vllm/blob/76ed5340f0ec0e481593ea1a94459b4b55136a4f/vllm/model_executor/layers/linear.py#L138

**Comment by IEI-mjx** [comment]
Sure, the Attention layer was defined like this:
```
class YuanAttention(nn.Module):

    def __init__(
        self,
        config: YuanConfig,
        hidden_size: int,
        attention_projection_size: int,
        num_heads: int,
        num_kv_heads: int,
        attn_head_size=None,
        rope_theta: float = 10000,
        rope_scaling: Optional[Dict[str, Any]] = None,
        max_position_embeddings: int = 8192,
        linear_method: Optional[LinearMethodBase] = None,
        bias: bool = False,
        sliding_window: Optional[int] = None,
        quant_config: Optional[QuantizationConfig] = None, # Quant Use
        #cache_config: Optional[CacheConfig] = None,
    ) -> None:
        super().__init__()
        self.config = config
        self.hidden_size = hidden_size
        self.tp_size = get_tensor_model_parallel_world_size()
        self.total_num_heads = num_heads
        self.lf_conv2d_group = config.lf_conv2d_group
        self.lf_conv2d_num_pad = config.lf_conv2d_num_pad

        self.attn_head_size = attention_projection_size // num_heads if attn_head_size is None else attn_head_size
        assert self.total_num_heads % self.tp_size == 0
        self.num_heads = self.total_num_heads // self.tp_size
        self.total_num_kv_heads = num_kv_heads
        if self.total_num_kv_heads >= self.tp_size:
            assert self.total_num_kv_heads % self.tp_size == 0
        else:
            assert self.tp_size % self.total_num_kv_heads == 0
        self.num_kv_heads = max(1, self.total_num_kv_heads // self.tp_size)
        self.head_dim = hidden_size // self.total_num_heads
        self.q_size = self.num_heads * self.head_dim
        self.kv_size = self.num_kv_heads * self.head_dim
        self.scaling = self.head_dim**-0.5
        self.rope_theta = rope_theta

        self.eps = 1e-6
        self.get_query_key = ColumnParallelLinear(
            hidden_size,
            2 * attention_projection_size,
            bias=False,
            quant_config=quant_config,
        )
        self.v_proj = ColumnParallelLinear(
            hidden_size,
            attention_projection_size,
            bias=False,
            quant_config=quant_config,
        )
        self.o_proj = RowParallelLinear(
            attention_projection_size,
            hidden_size,
            bias=False,
            quant_config=quant_config,
        )

        self.model_type = getattr(config, 'model_type', 'yuan')
        self.lf_gate = LocalizedFiltering(self.config, self.hidden_size, self.lf_conv2d_group, self.lf_conv2d_num_pad)
        self.attn = Attention(self.num_kv_heads,
                              self.attn_head_size,
                              self.scaling,
                              num_kv_heads=self.num_kv_heads,
                              #cache_config=cache_config,
                              quant_config=quant_config,
                              #sliding_window=sliding_window
                              )

    def forward(
        self,
        positions: torch.Tensor,
        hidden_states: torch.Tensor,
        rotary_pos_emb: torch.Tensor,
        kv_cache: torch.Tensor,
        lf1_cache: torch.Tensor,
        lf2_cache: torch.Tensor,
        attn_metadata: AttentionMetadata,
        use_yarn: bool=False,
        yarn_scale_factor: float=1.0,
        attn_factor: float=1.0,
    ) -> torch.Tensor:
        v, _ = self.v_proj(hidden_states)
        hidden_states, lf1, lf2 = self.lf_gate(hidden_states, lf1_cache, lf2_cache, attn_metadata)
        #v = v.view(*v.shape[:-1], self.num_heads, self.attn_head_size)
        qk, _ = self.get_query_key(hidden_states)
        qk = qk.view(*qk.shape[:-1], self.num_heads, int(qk.shape[-1] // self.num_heads))
        (q, k) = torch.chunk(qk, 2, dim=-1)
        q = apply_rotary_pos_emb(q, rotary_pos_emb, positions, use_yarn, yarn_scale_factor, attn_factor, attn_metadata)
        k = apply_rotary_pos_emb(k, rotary_pos_emb, positions, use_yarn, yarn_scale_factor, attn_factor, attn_metadata)
        #v = v.view(*v.shape[:-2], self.num_heads * self.attn_head_size)
        q = q.view(-1, self.num_heads * self.attn_head_size)
        k = k.view(-1, self.num_heads * self.attn_head_size)

        attn_output = self.attn(q, k, v, kv_cache, attn_metadata)
        output, _ = self.o_proj(attn_output)
        return output, lf1, lf2
```
 All of the linear layers in Attention are defined using the LinearBase layers, and even the `v_proj` and `o_proj` layer names contain no `weight_scales`.

**Comment by dsikka** [comment]
Can you ensure that this line is being called? https://github.com/vllm-project/vllm/blob/76ed5340f0ec0e481593ea1a94459b4b55136a4f/vllm/model_executor/layers/linear.py#L304
`quant_method` should be `CompressedTensorsLinearMethod` for quantized layers

**Comment by IEI-mjx** [comment]
Thank you for your reply,
I have fixed this issue! After adding the `**cache_config**`, which was previously missing from the script, the `weight_scale` values now appear correctly.
I am wondering if the parameters `quant_config` and `cache_config `should be used simultaneously？

**Comment by IEI-mjx** [comment]
Sorry for the additional question.
I have successfully run our FP8 model with `tp=1`. But when set `tp>1`, the output is poor. But the inference on unquantized bf16 model is great.
To address this issue, I tried changing the definitions of `w1` and `w2` in `ExpertLayer` from `Column/RowParallelLinear` to `ReplicatedLinear`.
While this change normalized the output, the performance scores on the GSM8K and HumanEval benchmarks dropped by 80%.
![image](https://github.com/user-attachments/assets/f771538e-c253-4e90-bd55-311a928182f4)
The only difference between fp8 and bf16 model is `quant_method.apply()`. Could you provide some guidance on how to handle this?
